src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -0,0 +1,524 @@
+{
+  NS_LOG_FUNCTION (this << minAllocation << maxAllocation);
+  return ((minAllocation + maxAllocation) / 2);
+}
+
+StatusCode
+PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info)
+{
+  NS_LOG_FUNCTION (this);
+
+  StatusCode status;
+  uint32_t allocDuration;
+  if (info.GetAllocationFormat () == ISOCHRONOUS)
+    {
+      allocDuration = GetAllocationDuration (dmgTspec.GetMinimumAllocation (), dmgTspec.GetMaximumAllocation ());
+    }
+  else if (info.GetAllocationFormat () == ASYNCHRONOUS)
+    {
+      // for asynchronous allocations, the Maximum Allocation field is reserved (IEEE 802.11ad 8.4.2.136)
+      allocDuration = dmgTspec.GetMinimumAllocation ();
+    }
+  else
+    {
+      NS_FATAL_ERROR ("Allocation Format not supported");
+    }
+
+  // Implementation of an admission policy for newly received requests. 
+  uint16_t allocPeriod = dmgTspec.GetAllocationPeriod ();
+  if (allocPeriod != 0)
+    {
+      // Proceed with the allocation of periodic SPs 
+      bool isMultiple = dmgTspec.IsAllocationPeriodMultipleBI ();
+      if (isMultiple)
+        {
+          NS_FATAL_ERROR ("Multiple BI periodicity is not supported yet.");
+        }
+
+      // spInterval is going to be passed to AddAllocationPeriod to specify the 
+      // distance between consecutive periodic SPs
+      uint32_t spInterval = uint32_t (m_biDuration.GetMicroSeconds () / allocPeriod);
+
+      NS_LOG_DEBUG ("Allocation Period " << uint32_t (allocPeriod) << " AllocDuration " << allocDuration << " Multiple " << isMultiple);
+      NS_LOG_DEBUG ("Schedule one SP every " << spInterval);
+
+      uint32_t startPeriodicAllocation = m_availableSlots[0].first;
+      uint32_t startFirstAllocation = startPeriodicAllocation;
+      uint32_t endAlloc;
+      uint8_t counter = 0;
+
+      uint8_t blocks = GetAvailableBlocks (allocDuration, spInterval);
+
+      if (blocks > 1)
+        {
+          while (counter < blocks)
+            {
+              NS_LOG_DEBUG ("Reserve from " << startPeriodicAllocation << " for " << allocDuration);
+
+              endAlloc = startPeriodicAllocation + allocDuration + m_guardTime;
+              UpdateAvailableSlots (startPeriodicAllocation, endAlloc);
+
+              startPeriodicAllocation += spInterval;
+              counter++;
+
+              // update remaining DTI time for consistency
+              m_remainingDtiTime -= (allocDuration + m_guardTime);
+            }
+
+          AddAllocationPeriod (info.GetAllocationID (), info.GetAllocationType (), info.IsPseudoStatic (),
+                               sourceAid, info.GetDestinationAid (),
+                               startFirstAllocation, allocDuration, spInterval, blocks);
+          status.SetSuccess ();
+        }
+      else
+        {
+          // if we cannot guarantee AT LEAST TWO periodic SPs, the request is rejected.
```

Comment:
This is also a good point and yes, we discussed about this offline. 
I will implement also this check, as you suggest :)

Reply:
In addition, should we check the number of blocks returned with the minimum allocation field in case of Isochronous traffic? If it's not possible to have more than one block again then reject